### PR TITLE
TAJO-1553: Improve broadcast join planning

### PR DESCRIPTION
**Pull request by jihoonson**
Sorry for the large patch, but most changes are related to unit tests.

In this patch, I've added ```BroadcastJoinRule``` as a new ```GlobalPlanRewriteRule```.
BroadcastJoinRule converts repartition join plan into broadcast join plan. To describe the broadcast join rules, we have to define the ```broadcastable``` property for a relation as follows.

_Broadcastable relation:_ A relation is broadcastable when its size is smaller than a given threshold.

And I've assumed that if every input of an execution block is broadcastable, the output of the execution block is also broadcastable.

Finally, here are the rules to convert repartition join into broadcast join.

* Given an EB containing a join and its child EBs, those EBs can be merged into a single EB if at least one child EB's output is broadcastable.
* Given a user-defined threshold, the total size of broadcast relations of an EB cannot exceed such threshold.
 * After merging EBs according to the first rule, the result EB may not satisfy the second rule. In this case, enforce repartition join for large relations to satisfy the second rule.
* Preserved-row relations cannot be broadcasted to avoid duplicated results. That is, full outer join cannot be executed with broadcast join.
 * Here is brief backgrounds for this rule. Data of preserved-row relations will be appeared in the join result regardless of join conditions. If multiple tasks execute outer join with broadcasted preserved-row relations, they emit duplicates results.
 * Even though a single task can execute outer join when every input is broadcastable, broadcast join is not allowed if one of input relation consists of multiple files.

[Review on Reviewable](https://reviewable.io/reviews/apache/tajo/583)

### CODE REVIEW

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java [hunk 1]
```diff
@@ -39,8 +40,38 @@
...
+   * 3) full outer join
+   *
+   *        full outer join
+   *          /        \
+   * null-supplying  preserved-row
+   * preserved-row   null-supplying
+   *
+   * The null-supplying and preserved-row flags are used to find which relations will be broadcasted.
+   */
+  protected transient boolean nullSuppllying = false;
```
Comment (hyunsik) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java]: 'transient' seems to be not necessary.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java [hunk 2]
```diff
@@ -120,25 +171,49 @@ public boolean hasUnion() {
...
-    broadcasted.add(tableName);
-    enforcer.addBroadcast(tableName);
+  public boolean isUnionOnly() {
+    return isUnionOnly;
   }
 
-  public void removeBroadcastTable(String tableName) {
-    broadcasted.remove(tableName);
-    enforcer.removeBroadcast(tableName);
+  public void addBroadcastRelation(ScanNode relationNode) {
```
Comment (hyunsik) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java]: any reason to change the suffix from table to relation?

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java (hunk 1 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java]: Thanks. I'll remove.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java (hunk 2 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/ExecutionBlock.java]: I used the term 'broadcastable'. If this term is used with 'table' together, i.e., broadcastableTable, it makes readers confused.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/enforce/Enforcer.java [hunk 3]
```diff
@@ -86,6 +86,23 @@ public void addSortedInput(String tableName, SortSpec[] sortSpecs) {
     TUtil.putToNestedList(properties, builder.getType(), builder.build());
   }
 
+  public void removeSortedInput(String tableName) {
```
Comment (hyunsik) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/enforce/Enforcer.java]: It seems to be not used in this patch.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/rewriter/rules/BroadcastJoinRule.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    plan.accept(plan.getRoot().getId(), planBuilder);
+    plan.accept(plan.getRoot().getId(), planFinalizer);
+    return plan;
+  }
+
+  private static class RelationSizeComparator implements Comparator<ScanNode> {
+
+    @Override
+    public int compare(ScanNode o1, ScanNode o2) {
+      return (int) (GlobalPlanRewriteUtil.getTableVolume(o1) - GlobalPlanRewriteUtil.getTableVolume(o2));
```
Comment (hyunsik) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/rewriter/rules/BroadcastJoinRule.java]: It seems to cause overflow when the difference of two table volumes are more than the range of integer value.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/enforce/Enforcer.java (hunk 3 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/enforce/Enforcer.java]: Removed.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/global/rewriter/rules/BroadcastJoinRule.java (hunk 4 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/global/rewriter/rules/BroadcastJoinRule.java]: Improved to consider overflow.

**Comment by jihoonson** [comment]
I focused on improving the broadcast join planning in this patch. I evaluated performance with an example query which Tajo cannot fully exploit the broadcast join opportunity. 
### Query
```
default> select r_name, case when l_shipmode is null then 'N/O' else l_shipmode end as s1 from region inner join ( select * from lineitem left outer join nation on l_suppkey = n_nationkey ) t on n_regionkey = r_regionkey
```
### Data
* Input: TPC-H scale factor of 100
* Result: 14484 rows (185.7 KiB)

### Performance
* Before patch: 226.915 sec
* After patch: 49.647 sec

### Query plan
##### Before patch
```
-------------------------------------------------------------------------------
Execution Block Graph (TERMINAL - eb_1432720176114_0001_000006)
-------------------------------------------------------------------------------
|-eb_1432720176114_0001_000006
   |-eb_1432720176114_0001_000005
      |-eb_1432720176114_0001_000004
      |-eb_1432720176114_0001_000001
-------------------------------------------------------------------------------
Order of Execution
-------------------------------------------------------------------------------
1: eb_1432720176114_0001_000001
2: eb_1432720176114_0001_000004
3: eb_1432720176114_0001_000005
4: eb_1432720176114_0001_000006
-------------------------------------------------------------------------------

=======================================================
Block Id: eb_1432720176114_0001_000001 [LEAF]
=======================================================

[Outgoing]
[q_1432720176114_0001] 1 => 5 (type=HASH_SHUFFLE, key=tpch100.region.r_regionkey (INT8), num=1186)

SCAN(0) on tpch100.region
  => target list: tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)
  => out schema: {(2) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)}
  => in schema: {(3) tpch100.region.r_regionkey (INT8), tpch100.region.r_name (TEXT), tpch100.region.r_comment (TEXT)}

=======================================================
Block Id: eb_1432720176114_0001_000004 [LEAF]
=======================================================

[Outgoing]
[q_1432720176114_0001] 4 => 5 (type=HASH_SHUFFLE, key=tpch100.t.n_regionkey (INT8), num=1186)

[Enforcers]
 0: type=Broadcast, tables=tpch100.nation

TABLE_SUBQUERY(5) as tpch100.t
  => Targets: CASE WHEN tpch100.t.l_shipmode (TEXT) IS NULL THEN N/O ELSE tpch100.t.l_shipmode (TEXT) END as s1, tpch100.t.n_regionkey (INT8)
  => out schema: {(2) s1 (TEXT), tpch100.t.n_regionkey (INT8)}
  => in  schema: {(20) tpch100.t.l_orderkey (INT8), tpch100.t.l_partkey (INT8), tpch100.t.l_suppkey (INT8), tpch100.t.l_linenumber (INT8), tpch100.t.l_quantity (FLOAT8), tpch100.t.l_extendedprice (FLOAT8), tpch100.t.l_discount (FLOAT8), tpch100.t.l_tax (FLOAT8), tpch100.t.l_returnflag (TEXT), tpch100.t.l_linestatus (TEXT), tpch100.t.l_shipdate (TEXT), tpch100.t.l_commitdate (TEXT), tpch100.t.l_receiptdate (TEXT), tpch100.t.l_shipinstruct (TEXT), tpch100.t.l_shipmode (TEXT), tpch100.t.l_comment (TEXT), tpch100.t.n_nationkey (INT8), tpch100.t.n_name (TEXT), tpch100.t.n_regionkey (INT8), tpch100.t.n_comment (TEXT)}
   PROJECTION(4)
     => Targets: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
     => out schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
     => in  schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
      JOIN(9)(LEFT_OUTER)
        => Join Cond: tpch100.lineitem.l_suppkey (INT8) = tpch100.nation.n_nationkey (INT8)
        => target list: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
        => out schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
        => in schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
         SCAN(2) on tpch100.nation
           => target list: tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
           => out schema: {(4) tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
           => in schema: {(4) tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
         SCAN(1) on tpch100.lineitem
           => target list: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)
           => out schema: {(16) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)}
           => in schema: {(16) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)}

=======================================================
Block Id: eb_1432720176114_0001_000005 [ROOT]
=======================================================

[Incoming]
[q_1432720176114_0001] 1 => 5 (type=HASH_SHUFFLE, key=tpch100.region.r_regionkey (INT8), num=1186)
[q_1432720176114_0001] 4 => 5 (type=HASH_SHUFFLE, key=tpch100.t.n_regionkey (INT8), num=1186)

JOIN(10)(INNER)
  => Join Cond: tpch100.t.n_regionkey (INT8) = tpch100.region.r_regionkey (INT8)
  => target list: tpch100.region.r_name (TEXT), s1 (TEXT)
  => out schema: {(2) tpch100.region.r_name (TEXT), s1 (TEXT)}
  => in schema: {(4) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8), s1 (TEXT), tpch100.t.n_regionkey (INT8)}
   SCAN(13) on eb_1432720176114_0001_000004
     => out schema: {(2) s1 (TEXT), tpch100.t.n_regionkey (INT8)}
     => in schema: {(2) s1 (TEXT), tpch100.t.n_regionkey (INT8)}
   SCAN(12) on eb_1432720176114_0001_000001
     => out schema: {(2) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)}
     => in schema: {(2) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)}

=======================================================
Block Id: eb_1432720176114_0001_000006 [TERMINAL]
=======================================================
```
##### After patch
```
-------------------------------------------------------------------------------
Execution Block Graph (TERMINAL - eb_1432719753330_0002_000006)
-------------------------------------------------------------------------------
|-eb_1432719753330_0002_000006
   |-eb_1432719753330_0002_000005
-------------------------------------------------------------------------------
Order of Execution
-------------------------------------------------------------------------------
1: eb_1432719753330_0002_000005
2: eb_1432719753330_0002_000006
-------------------------------------------------------------------------------

=======================================================
Block Id: eb_1432719753330_0002_000005 [ROOT]
=======================================================

[Enforcers]
 0: type=Broadcast, tables=tpch100.region
 1: type=Broadcast, tables=tpch100.nation

JOIN(10)(INNER)
  => Join Cond: tpch100.t.n_regionkey (INT8) = tpch100.region.r_regionkey (INT8)
  => target list: tpch100.region.r_name (TEXT), s1 (TEXT)
  => out schema: {(2) tpch100.region.r_name (TEXT), s1 (TEXT)}
  => in schema: {(4) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8), s1 (TEXT), tpch100.t.n_regionkey (INT8)}
   TABLE_SUBQUERY(5) as tpch100.t
     => Targets: CASE WHEN tpch100.t.l_shipmode (TEXT) IS NULL THEN N/O ELSE tpch100.t.l_shipmode (TEXT) END as s1, tpch100.t.n_regionkey (INT8)
     => out schema: {(2) s1 (TEXT), tpch100.t.n_regionkey (INT8)}
     => in  schema: {(20) tpch100.t.l_orderkey (INT8), tpch100.t.l_partkey (INT8), tpch100.t.l_suppkey (INT8), tpch100.t.l_linenumber (INT8), tpch100.t.l_quantity (FLOAT8), tpch100.t.l_extendedprice (FLOAT8), tpch100.t.l_discount (FLOAT8), tpch100.t.l_tax (FLOAT8), tpch100.t.l_returnflag (TEXT), tpch100.t.l_linestatus (TEXT), tpch100.t.l_shipdate (TEXT), tpch100.t.l_commitdate (TEXT), tpch100.t.l_receiptdate (TEXT), tpch100.t.l_shipinstruct (TEXT), tpch100.t.l_shipmode (TEXT), tpch100.t.l_comment (TEXT), tpch100.t.n_nationkey (INT8), tpch100.t.n_name (TEXT), tpch100.t.n_regionkey (INT8), tpch100.t.n_comment (TEXT)}
      PROJECTION(4)
        => Targets: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
        => out schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
        => in  schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
         JOIN(9)(LEFT_OUTER)
           => Join Cond: tpch100.lineitem.l_suppkey (INT8) = tpch100.nation.n_nationkey (INT8)
           => target list: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
           => out schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
           => in schema: {(20) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT), tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
            SCAN(2) on tpch100.nation
              => target list: tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)
              => out schema: {(4) tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
              => in schema: {(4) tpch100.nation.n_nationkey (INT8), tpch100.nation.n_name (TEXT), tpch100.nation.n_regionkey (INT8), tpch100.nation.n_comment (TEXT)}
            SCAN(1) on tpch100.lineitem
              => target list: tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)
              => out schema: {(16) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)}
              => in schema: {(16) tpch100.lineitem.l_orderkey (INT8), tpch100.lineitem.l_partkey (INT8), tpch100.lineitem.l_suppkey (INT8), tpch100.lineitem.l_linenumber (INT8), tpch100.lineitem.l_quantity (FLOAT8), tpch100.lineitem.l_extendedprice (FLOAT8), tpch100.lineitem.l_discount (FLOAT8), tpch100.lineitem.l_tax (FLOAT8), tpch100.lineitem.l_returnflag (TEXT), tpch100.lineitem.l_linestatus (TEXT), tpch100.lineitem.l_shipdate (TEXT), tpch100.lineitem.l_commitdate (TEXT), tpch100.lineitem.l_receiptdate (TEXT), tpch100.lineitem.l_shipinstruct (TEXT), tpch100.lineitem.l_shipmode (TEXT), tpch100.lineitem.l_comment (TEXT)}
   SCAN(0) on tpch100.region
     => target list: tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)
     => out schema: {(2) tpch100.region.r_name (TEXT), tpch100.region.r_regionkey (INT8)}
     => in schema: {(3) tpch100.region.r_regionkey (INT8), tpch100.region.r_name (TEXT), tpch100.region.r_comment (TEXT)}

=======================================================
Block Id: eb_1432719753330_0002_000006 [TERMINAL]
=======================================================
```

**Comment by hyunsik** [comment]
+1 The patch looks good to me. The result looks significant!

**Pull request closed without merging**